modmail/extensions/utils/error_handler.py

```diff
@@ -0,0 +1,195 @@
+    @staticmethod
+    def error_embed(message: str, title: str = None) -> discord.Embed:
+        """Create an error embed with an error colour and reason and return it."""
+        return discord.Embed(message, colour=ERROR_COLOUR, title=title or "Error Occurred")
+
+    @staticmethod
+    def get_title_from_name(error: typing.Union[Exception, str]) -> str:
+        """
+        Return a message dervived from the exception class name.
+
+        Eg NSFWChannelRequired returns NSFW Channel Required
+        """
+        if not isinstance(error, str):
+            error = error.__class__.__name__
+        return re.sub(ERROR_TITLE_REGEX, r" \1", error)
+
+    @staticmethod
+    def _reset_command_cooldown(ctx: commands.Context) -> bool:
+        if return_value := ctx.command.is_on_cooldown(ctx):
+            ctx.command.reset_cooldown(ctx)
+        return return_value
+
+    async def handle_user_input_error(
+        self,
+        ctx: commands.Context,
+        error: commands.UserInputError,
+        reset_cooldown: bool = True,
+    ) -> discord.Embed:
+        """Handling deferred from main error handler to handle UserInputErrors."""
+        if reset_cooldown:
+            self._reset_command_cooldown(ctx)
+        embed = None
+        msg = None
+        title = "User Input Error"
+        if isinstance(error, commands.BadUnionArgument):
+            msg = self.get_title_from_name(str(error))
+            title = self.get_title_from_name(error)
+        else:
+            title = self.get_title_from_name(error)
+        return embed or self.error_embed(msg or str(error), title=title)
+
+    async def handle_bot_missing_perms(
+        self, ctx: commands.Context, error: commands.BotMissingPermissions
+    ) -> bool:
+        """Handles bot missing permissing by dming the user if they have a permission which may be able to fix this."""  # noqa: E501
+        embed = self.error_embed(str(error))
+        try:
+            await ctx.send(embeds=[embed])
+        except discord.Forbidden:
+            logger.error(f"Unable to send an error message to {ctx.channel!s}")
+            if ANY_DEV_MODE:
+                # non-general permissions
+                perms = discord.Permissions(
+                    administrator=True,
+                    manage_threads=True,
+                    manage_roles=True,
+                    manage_channels=True,
+                )
+                if perms.value & ctx.channel.permissions_for(ctx.author).value:
+                    logger.info(
+                        f"Attempting to dm {ctx.author} since they have a permission which may be able "
+                        "to give the bot send message permissions."
+                    )
+                    try:
+                        await ctx.author.send(embeds=[embed])
+                    except discord.Forbidden:
+                        logger.notice("Also encountered an error when trying to reply in dms.")
+                        return False
```

Comment:
I love this feature. Needs to be behind a configuration toggle, though.

Reply:
Configuration system is in a different pr, GH-75, so I'll add a constant for now.

Reply:
Actually, this was behind a partial config toggle. If the bot is in a developer mode it will the dm the user, but was not turned on for main users.

Reply:
I'll switch this from the developer default to the configuration system once it is merged.